sections/loans/components/ActionBox/components/Balance.tsx

```diff
@@ -0,0 +1,171 @@
+		<ERC20 {...{ asset, onSetMaxAmount }} />
+	);
+};
+
+export default Balance;
+
+type ETHProps = {
+	onSetMaxAmount?: (amount: string) => void;
+};
+
+const ETH: React.FC<ETHProps> = ({ onSetMaxAmount }) => {
+	const { t } = useTranslation();
+	const { provider, signer } = Connector.useContainer();
+	const [balance, setBalance] = React.useState(ethers.BigNumber.from('0'));
+
+	const handleSetMaxAmount = () => {
+		if (onSetMaxAmount && balance) {
+			onSetMaxAmount(ethers.utils.formatUnits(balance, 18));
+		}
+	};
+
+	React.useEffect(() => {
+		if (!signer) return;
+
+		let isMounted = true;
+		const unsubs: Array<any> = [() => (isMounted = false)];
+
+		const onSetBalance = async () => {
+			const balance = await signer.getBalance();
+			if (isMounted) setBalance(balance);
+		};
+
+		const subscribe = () => {
+			if (provider) {
+				const newBlockEvent = 'block';
+				provider.on(newBlockEvent, onSetBalance);
+				unsubs.push(() => provider.off(newBlockEvent, onSetBalance));
+			}
+		};
+
+		onSetBalance();
+		subscribe();
+		return () => {
+			unsubs.forEach((unsub) => unsub());
+		};
+	}, [signer, provider]);
+
+	return (
+		balance && (
+			<Container>
+				{t('balance.input-label')} {formatUnits(balance, 18, 2)}{' '}
+				{!onSetMaxAmount ? null : <MaxButton onClick={handleSetMaxAmount} />}
+			</Container>
+		)
+	);
+};
+
+type ERC20Props = {
+	asset: string;
+	onSetMaxAmount?: (amount: string) => void;
+};
+
+const ERC20: React.FC<ERC20Props> = ({ asset, onSetMaxAmount }) => {
+	const { t } = useTranslation();
+	const address = useRecoilValue(walletAddressState);
+	const [balance, setBalance] = React.useState<ethers.BigNumber>(ethers.BigNumber.from('0'));
+	const [decimals, setDecimals] = React.useState<number>(0);
+	const { renBTCContract } = useLoans();
+
+	const handleSetMaxAmount = () => {
+		if (onSetMaxAmount && balance && decimals) {
+			onSetMaxAmount(ethers.utils.formatUnits(balance, decimals));
+		}
+	};
+
+	const contract = React.useMemo(() => {
+		const {
+			contracts: { ProxysBTC: sBTC, ProxysETH: sETH, ProxyERC20sUSD: sUSD },
+		} = synthetix.js!;
+		const tokens: Record<string, ethers.Contract> = { sBTC, sETH, sUSD, renBTC: renBTCContract! };
+		return tokens[asset];
+	}, [asset, renBTCContract]);
+
+	React.useEffect(() => {
+		if (!(contract && address)) return;
+
+		let isMounted = true;
+		const unsubs: Array<any> = [() => (isMounted = false)];
+
+		const loadBalance = async () => {
+			const [decimals, balance] = await Promise.all([
+				contract.decimals(),
+				contract.balanceOf(address),
+			]);
+			if (isMounted) {
+				setDecimals(decimals);
+				setBalance(balance);
+			}
+		};
+
+		const subscribe = () => {
+			const transferEvent = contract.filters.Transfer();
+			const onBalanceChange = async (from: string, to: string) => {
+				if (from === address || to === address) {
+					if (isMounted) setBalance(await contract.balanceOf(address));
+				}
+			};
+
+			contract.on(transferEvent, onBalanceChange);
```

Comment:
Events are really bad with ethers.js. They keep polling the infura node, every 10 secs or so.Any way we can use the graph for this?

Reply:
Happy to help you setting this up if needed